Sources/Algorithms/Keyed.swift

```diff
@@ -0,0 +1,70 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension Sequence {
+  /// Creates a new Dictionary from the elements of `self`, keyed by the
+  /// results returned by the given `keyForValue` closure. As the dictionary is
+  /// built, the initializer calls the `combine` closure with the current and
+  /// new values for any duplicate keys. Pass a closure as `combine` that
+  /// returns the value to use in the resulting dictionary: The closure can
+  /// choose between the two values, combine them to produce a new value, or
+  /// even throw an error.
+  ///
+  /// If no `combine` closure is provided, deriving the same duplicate key for
+  /// more than one element of self results in a runtime error.
+  ///
+  /// - Parameters:
+  ///   - keyForValue:  A closure that returns a key for each element in
+  ///     `self`.
+  ///   - combine: A closure that is called with the values for any duplicate
+  ///     keys that are encountered. The closure returns the desired value for
+  ///     the final dictionary.
+  @inlinable
+  public func keyed<Key>(
+    by keyForValue: (Element) throws -> Key,
+    uniquingKeysWith combine: ((Key, Element, Element) throws -> Element)? = nil
```

Comment:
Is `uniquingKeysWith` still the right naming? After all, it's not the keys being uniqued, it's the values.

Reply:
Good point.  It does follow convention from the `Dictionary` initialiser, though.

Maybe `resolvingConflictsWith`?

Technically this is equivalent to first grouping by the keys then reducing the resulting dictionary values… so maybe `reducingWith`?  In a way that's nicer because it reminds us that you don't _have_ to return one or other of the given values - you can technically return any value of the same type, such as the sum of both values, or the largest or smallest, etc.

There's also no way, with this design, to preserve both elements but under different keys.  But I'm not sure that needs to be supported… in such cases, perhaps `group(by:)` is sufficient?

Reply:
> Maybe `resolvingConflictsWith`?

I'm digging this. "Resolving" is clear enough in context, while succinct and not overly prescriptive as to what the closure's resolution logic should be

> There's also no way, with this design, to preserve both elements but under different keys. But I'm not sure that needs to be supported… in such cases, perhaps group(by:) is sufficient?

Agree.

Reply:
Is this something you'd like to commit? You deserve to have your name on your findings :)